File: Program_4/4(a).c
#include <stdio.h>
#include <time.h>

void fibonachi_by_recursive(int n, int x, int y) {
    if (n == 0)
        return;
    if (n == 1) {
        // printf("%d", x);
        return;
    }
    // printf("%d %d ", x, y);
    x = x + y;
    y = y + x;
    fibonachi_by_recursive(n - 2, x, y);
}

int main() {
    int sizes[] = {100, 1000, 5000, 10000};
    int num_sizes = 4;
    int repetitions = 10000;  

    for (int i = 0; i < num_sizes; i++) {
        int n = sizes[i];

        clock_t start = clock();
        for (int j = 0; j < repetitions; j++) {
            fibonachi_by_recursive(n, 0, 1);
        }
        clock_t end = clock();

        double time = ((double)(end - start)) / CLOCKS_PER_SEC / repetitions;
        printf("for size=%d time=%f\n",n,time);  
    }

    return 0;
}

File: Program_4/4(a).py
import time
import matplotlib.pyplot as plt

def fibonachi_by_recursive(n, x=0, y=1):
    if n == 0:
        return
    if n == 1:
        # print(x, end=" ")
        return
    # print(x, y, end=" ")
    x = x + y
    y = y + x
    fibonachi_by_recursive(n - 2, x, y)

sizes = [10, 100, 500, 1000]
times = []

for n in sizes:
    start = time.perf_counter()
    fibonachi_by_recursive(n)
    end = time.perf_counter()
    elapsed = end - start
    times.append(elapsed)
    print(f"For input size {n}, time taken = {elapsed:.6f} sec")

plt.plot(sizes, times, marker='o', label="Recursive Sequence Generator")
plt.xlabel("Input Size (n)")
plt.ylabel("Time Taken (seconds)")
plt.title("Recursive Fibonacci Sequence Generator Performance")
plt.legend()
plt.grid(True)
plt.show()
